staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -1186,6 +1190,36 @@ func schemaIsNonStructural(schema *apiextensions.JSONSchemaProps) bool {
 	return len(structuralschema.ValidateStructural(nil, ss)) > 0
 }
 
+// requirePrunedDefaults returns true if the old CRD spec had no unpruned defaults.
+func requirePrunedDefaults(spec *apiextensions.CustomResourceDefinitionSpec) bool {
+	if spec.Validation != nil && schemaIsNonStructuralOrHasPrunedDefaults(spec.Validation.OpenAPIV3Schema) {
+		return true
+	}
+	for _, v := range spec.Versions {
+		if v.Schema != nil && schemaIsNonStructuralOrHasPrunedDefaults(v.Schema.OpenAPIV3Schema) {
+			return true
```

Comment:
This breaks if another later version in spec.Versions has unpruned defaults

Reply:
I think this actually needs to flip what it's checking:
If:
* schema has defaults AND
* schema is non-structural OR schema has unpruned defaults

Then return *false*

Reply:
Same for the check of the top-level validation

Reply:
Yep, wrong copy&paste from above. Fixed.

Reply:
Now we only accept unpruned defaults if we have had unpruned defaults before. If it has been non-structual, that's not enough.